### PR TITLE
[FLINK-15616][python] Move boot error messages from python-udf-boot.log to taskmanager's log file

### PR DESCRIPTION
**Pull request by hequn8128**
## What is the purpose of the change

Previously, the boot error messages are printed in the log file under FLINK_LOG_DIR, i.e., "$FLINK_LOG_DIR/flink-$USER-python-udf-boot-$HOSTNAME.log". This additional file is very hard to locate for users.

This pull request also prints the error messages into the taskmanager log file to make it more user-friendly. And throws an exception when there is an error during boot.

## Brief change log

  - Put the boot log file in the taskmanager's temp directly.
  - Prints the error messages into the taskmanager log file to make it more user-friendly.
  - Throws an exception when there is an error during boot.

## Verifying this change

This change is already covered by existing tests, such as `ProcessPythonEnvironmentManagerTest.testSetLogDirectory()`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java [hunk 1]
```diff
@@ -176,10 +179,29 @@ public void open() throws Exception {
...
 	}
 
+	/**
+	 * To make the error messages more user friendly, read the boot logs from the temp log file and
+	 * throw an exception directly.
+	 */
+	private StageBundleFactory createStageBundleFactory() throws Exception {
+		try {
+			return jobBundleFactory.forStage(createExecutableStage());
+		} catch (Exception e) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java]: catch Throwable?

File: flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java [hunk 2]
```diff
@@ -176,10 +179,29 @@ public void open() throws Exception {
...
+			byte[] output =
+				Files.readAllBytes(
+					new File(((ProcessPythonEnvironmentManager) environmentManager).getBaseDirectory()
+						+ "/flink-python-udf-boot.log").toPath());
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java]: What about changing `/flink-python-udf-boot.log` to `File.separator + "flink-python-udf-boot.log`?

File: flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java [hunk 3]
```diff
@@ -176,10 +179,29 @@ public void open() throws Exception {
...
+	 * To make the error messages more user friendly, read the boot logs from the temp log file and
+	 * throw an exception directly.
+	 */
+	private StageBundleFactory createStageBundleFactory() throws Exception {
+		try {
+			return jobBundleFactory.forStage(createExecutableStage());
+		} catch (Exception e) {
+			byte[] output =
+				Files.readAllBytes(
+					new File(((ProcessPythonEnvironmentManager) environmentManager).getBaseDirectory()
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java]: What happens if it's not an instance of ProcessPythonEnvironmentManager?

File: flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java [hunk 4]
```diff
@@ -176,10 +179,29 @@ public void open() throws Exception {
...
+	/**
+	 * To make the error messages more user friendly, read the boot logs from the temp log file and
+	 * throw an exception directly.
+	 */
+	private StageBundleFactory createStageBundleFactory() throws Exception {
+		try {
+			return jobBundleFactory.forStage(createExecutableStage());
+		} catch (Exception e) {
+			byte[] output =
+				Files.readAllBytes(
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java]: Should we check if the log file exists?

File: flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java (hunk 3 above)
Comment (hequn8128) [reply flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java]: How about adding the `getBaseDirectory` method into the interface?

**Comment by dianfu** [comment]
@hequn8128 Thanks a lot for the update. LGTM. +1 to merge.

**Comment by hequn8128** [comment]
@dianfu Thanks a lot for the review and good suggestions! Will merge this once travis passed.

**Pull request closed without merging**